src/metaswitch/clearwater/cluster_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,304 @@
+                        # there has been a snapshot between the get and the
+                        # watch. Just start the read again.
+                        _log.info("etcd index {} is invalid, retrying".format(
+                            result.modifiedIndex+1))
+                        self.read_from_etcd()
+
+                # Return if we're termiating.
+                if self._terminate_flag:
+                    return
+                else:
+                    cluster_view = self.parse_cluster_view(result.value)
+
+            # Save off the index of the result we're using for when we write
+            # back to etcd later.
+            self._index = result.modifiedIndex
+            self._last_cluster_view = cluster_view.copy()
+
+        except etcd.EtcdKeyError:
+            # If the key doesn't exist in etcd then there is currently no
+            # cluster.
+            self._index = None
+            self._last_cluster_view = None
+            pass
+        except Exception as e:
+            # Catch-all error handler (for invalid requests, timeouts, etc -
+            # unset all our state and start over.
+            _log.error("{} caught {!r} when trying to read with index {}"
+                       " - pause before retry".
+                       format(self._ip, e, self._index))
+            self._index = None
+            self._last_cluster_view = None
+            # Sleep briefly to avoid hammering a failed server
+            sleep(self.PAUSE_BEFORE_RETRY)
+
+        return cluster_view
+
+    # Write the new cluster view to etcd. We may be expecting to create the key
+    # for the first time.
+    def write_to_etcd(self, old_cluster_view, new_state, with_index=None):
+        cluster_view = old_cluster_view.copy()
+
+        # Update the cluster view based on new state information. If new_state
+        # is a string then it refers to the new state of the local node.
+        # Otherwise, it is an overall picture of the new cluster.
+        if isinstance(new_state, str):
+            cluster_view[self._ip] = new_state
+        elif isinstance(new_state, dict):
+            cluster_view = new_state
+
+        _log.debug("Writing state %s into etcd" % cluster_view)
+        json_data = json.dumps(cluster_view)
+
+        try:
+            if with_index:
+                self._client.write(self._key, json_data, prevIndex=with_index)
+            if self._index is None:
+                self._client.write(self._key, json_data, prevExist=False)
+            else:
+                self._client.write(self._key, json_data, prevIndex=self._index)
+
+                # We may have just successfully set the local node to
+                # WAITING_TO_LEAVE, in which case we no longer need the leaving
+                # flag.
+                self._leaving_flag = False
+        except ValueError:
+            _log.debug("Contention on etcd write")
```

Comment:
In principle a ValueError could be thrown for other reasons (e.g if we tried to call int("hello world") somewhere) but that would be a code bug which we should catch in UT. I don't think there's any way to catch the ValueError that etcd.Client.write throws without also catching that.

I don't think we should have a brief pause - we don't do this in other places where we spot a CAS failure and write again, like in memcached. (The reason we have pauses in other places in the code is where we might hit permanent errors - like etcd being down - whereas a CAS failure will resolve itself after two or three writes.)